File: plan/types/protocols/inbounds.ts
import type { Address, PortValue } from "../shared";
import type { StreamSettingsObject } from "../transports";
import type {
  InboundSettingsDokodemo,
  InboundSettingsHttp,
  InboundSettingsHysteria,
  InboundSettingsShadowsocks,
  InboundSettingsSocks,
  InboundSettingsTrojan,
  InboundSettingsTun,
  InboundSettingsTunnel,
  InboundSettingsVless,
  InboundSettingsVmess,
  InboundSettingsWireguard,
} from "./inbound-settings";

export interface SniffingObject {
  enabled?: boolean;
  destOverride?: Array<"http" | "tls" | "quic" | "fakedns">;
  metadataOnly?: boolean;
  domainsExcluded?: string[];
  ipsExcluded?: string[];
  routeOnly?: boolean;
}

interface BaseInboundObject {
  listen?: Address;
  port: PortValue;
  streamSettings?: StreamSettingsObject;
  tag?: string;
  sniffing?: SniffingObject;
}

export interface DokodemoInboundObject extends BaseInboundObject {
  protocol: "dokodemo-door";
  settings: InboundSettingsDokodemo;
}

export interface HttpInboundObject extends BaseInboundObject {
  protocol: "http";
  settings: InboundSettingsHttp;
}

export interface HysteriaInboundObject extends BaseInboundObject {
  protocol: "hysteria";
  settings: InboundSettingsHysteria;
}

export interface ShadowsocksInboundObject extends BaseInboundObject {
  protocol: "shadowsocks";
  settings: InboundSettingsShadowsocks;
}

export interface SocksInboundObject extends BaseInboundObject {
  protocol: "socks";
  settings: InboundSettingsSocks;
}

export interface TrojanInboundObject extends BaseInboundObject {
  protocol: "trojan";
  settings: InboundSettingsTrojan;
}

export interface TunInboundObject extends BaseInboundObject {
  protocol: "tun";
  settings: InboundSettingsTun;
}

export interface TunnelInboundObject extends BaseInboundObject {
  protocol: "tunnel";
  settings: InboundSettingsTunnel;
}

export interface VlessInboundObject extends BaseInboundObject {
  protocol: "vless";
  settings: InboundSettingsVless;
}

export interface VmessInboundObject extends BaseInboundObject {
  protocol: "vmess";
  settings: InboundSettingsVmess;
}

export interface WireguardInboundObject extends BaseInboundObject {
  protocol: "wireguard";
  settings: InboundSettingsWireguard;
}

export type InboundObject =
  | DokodemoInboundObject
  | HttpInboundObject
  | HysteriaInboundObject
  | ShadowsocksInboundObject
  | SocksInboundObject
  | TrojanInboundObject
  | TunInboundObject
  | TunnelInboundObject
  | VlessInboundObject
  | VmessInboundObject
  | WireguardInboundObject;

File: src/xray/parsing/protocols/inbounds.rs
use serde::{Deserialize, Deserializer, Serialize};
use serde_json::Value;

use super::inbound_settings::*;
use crate::xray::parsing::shared::{Address, PortValue};
use crate::xray::parsing::transports::StreamSettingsObject;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InboundObject {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub listen: Option<Address>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub port: Option<PortValue>,
    pub protocol: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub settings: Option<InboundSettings>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stream_settings: Option<StreamSettingsObject>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tag: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sniffing: Option<SniffingObject>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(untagged)]
pub enum InboundSettings {
    Dokodemo(InboundSettingsDokodemo),
    Http(InboundSettingsHttp),
    Hysteria(InboundSettingsHysteria),
    Shadowsocks(InboundSettingsShadowsocks),
    Socks(InboundSettingsSocks),
    Trojan(InboundSettingsTrojan),
    Tun(InboundSettingsTun),
    Tunnel(InboundSettingsTunnel),
    Vless(InboundSettingsVless),
    Vmess(InboundSettingsVmess),
    Wireguard(InboundSettingsWireguard),
    Unknown(Value),
}

impl InboundObject {
    pub(crate) fn has_unknown_protocol(&self) -> bool {
        !matches!(
            self.protocol.as_str(),
            "dokodemo-door"
                | "http"
                | "hysteria"
                | "shadowsocks"
                | "socks"
                | "mixed"
                | "trojan"
                | "tun"
                | "tunnel"
                | "vless"
                | "vmess"
                | "wireguard"
        )
    }
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct RawInboundObject {
    listen: Option<Address>,
    port: Option<PortValue>,
    protocol: String,
    settings: Option<Value>,
    stream_settings: Option<StreamSettingsObject>,
    tag: Option<String>,
    sniffing: Option<SniffingObject>,
}

impl<'de> Deserialize<'de> for InboundObject {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let raw = RawInboundObject::deserialize(deserializer)?;
        let settings = raw
            .settings
            .map(|value| {
                let parsed = match raw.protocol.as_str() {
                    "dokodemo-door" => serde_json::from_value(value).map(InboundSettings::Dokodemo),
                    "http" => serde_json::from_value(value).map(InboundSettings::Http),
                    "hysteria" => serde_json::from_value(value).map(InboundSettings::Hysteria),
                    "shadowsocks" => {
                        serde_json::from_value(value).map(InboundSettings::Shadowsocks)
                    }
                    "socks" | "mixed" => serde_json::from_value(value).map(InboundSettings::Socks),
                    "trojan" => serde_json::from_value(value).map(InboundSettings::Trojan),
                    "tun" => serde_json::from_value(value).map(InboundSettings::Tun),
                    "tunnel" => serde_json::from_value(value).map(InboundSettings::Tunnel),
                    "vless" => serde_json::from_value(value).map(InboundSettings::Vless),
                    "vmess" => serde_json::from_value(value).map(InboundSettings::Vmess),
                    "wireguard" => serde_json::from_value(value).map(InboundSettings::Wireguard),
                    _ => Ok(InboundSettings::Unknown(value)),
                };
                parsed.map_err(serde::de::Error::custom)
            })
            .transpose()?;

        Ok(Self {
            listen: raw.listen,
            port: raw.port,
            protocol: raw.protocol,
            settings,
            stream_settings: raw.stream_settings,
            tag: raw.tag,
            sniffing: raw.sniffing,
        })
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SniffingObject {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enabled: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dest_override: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata_only: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub domains_excluded: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ips_excluded: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub route_only: Option<bool>,
}
